tripy/tripy/backend/compiler_api.py

```diff
@@ -31,6 +32,91 @@
 from tripy.common.shape_bounds import ShapeBounds
 from tripy.frontend import Tensor, Trace
 from tripy.utils import json as json_utils
+from tripy.backend.mlir.utils import MLIRRuntimeClient
+
+
+@export.public_api(document_under="compiler")
+class Stream:
+    """
+    Represents a CUDA stream that can be used to manage concurrent operations.
+
+    This class is a wrapper around the underlying stream object, allowing management of CUDA streams.
+
+    .. code-block:: python
+        :linenos:
+        :caption: Use streams in execution
+
+        linear = tp.Linear(2, 3)
+        compiler = tp.Compiler(linear)
+
+        compiled_linear = compiler.compile(tp.InputInfo((2, 2), dtype=tp.float32))
+        with tp.Stream():
+            a = tp.ones((2, 2), dtype=tp.float32)
+            out = compiled_linear(a)
+
+        assert cp.array_equal(cp.from_dlpack(out), cp.from_dlpack(linear(a)))
+    """
+
+    _active_stream = None
+    _default_stream = None
+
+    def __init__(self):
+        self._stream = MLIRRuntimeClient().create_stream()
+        Stream._active_stream = self
+
+    @classmethod
+    def default_stream(cls):
+        """Get the default stream, create it if it doesn't exist."""
+        if cls._default_stream is None:
+            cls._default_stream = cls.__new__(cls)
+            cls._default_stream._stream = MLIRRuntimeClient().create_stream()
+            cls._active_stream = cls._default_stream
+
+        return cls._default_stream
+
+    def synchronize(self):
+        """Synchronize the stream, blocking until all operations in this stream are complete."""
+        self._stream.sync()
+
+    @classmethod
+    def get_current_stream(cls):
+        return cls._active_stream
+
+    def __enter__(self):
+        return self
+
+    def __exit__(self, exc_type, exc_value, traceback):
+        self.synchronize()
+        # TODO: expose methods from mlir-tensorrt python bindings to destroy stream
+        if self == Stream._active_stream:
+            Stream._active_stream = Stream._default_stream
+
+    def __eq__(self, other):
+        if not isinstance(other, Stream):
+            return False
+
+        if not (hasattr(self, "_stream") and hasattr(other, "_stream")):
+            return False
+
+        return self._stream == other._stream
+
+    def __repr__(self):
+        return f"<Stream(id={id(self)}, default={self == Stream._default_stream})>"
+
+    def __hash__(self):
```

Comment:
Do we need to define hash? If yes, id will not be reliable since different streams could have the same id over the lifetime of the program

Reply:
Not required anymore. Removed from latest iteration.